### PR TITLE
Synchronize access to Liquibase APIs.

### PR DESCRIPTION
**Pull request by kshakir**

### CODE REVIEW

File: database/migration/src/main/scala/cromwell/database/migration/liquibase/LiquibaseUtils.scala [hunk 1]
```diff
@@ -11,6 +11,7 @@ import liquibase.{Contexts, LabelExpression, Liquibase}
 import org.hsqldb.persist.HsqlDatabaseProperties
 
 object LiquibaseUtils {
+  private val mutex = new Object
```
Comment (Horneth) [inline database/migration/src/main/scala/cromwell/database/migration/liquibase/LiquibaseUtils.scala]: OOC what is the difference between this and synchronizing on `LiquibaseUtils` itself ?

File: database/migration/src/main/scala/cromwell/database/migration/liquibase/LiquibaseUtils.scala [hunk 2]
```diff
@@ -21,16 +22,18 @@ object LiquibaseUtils {
     * @param jdbcConnection A jdbc connection to the database.
     */
   def updateSchema(settings: LiquibaseSettings)(jdbcConnection: Connection): Unit = {
-    val liquibaseConnection = newConnection(jdbcConnection)
-    try {
-      val database = DatabaseFactory.getInstance.findCorrectDatabaseImplementation(liquibaseConnection)
-      database.setDatabaseChangeLogLockTableName(settings.databaseChangeLogLockTableName.toUpperCase)
-      database.setDatabaseChangeLogTableName(settings.databaseChangeLogTableName.toUpperCase)
+    mutex.synchronized {
```
Comment (aednichols) [inline database/migration/src/main/scala/cromwell/database/migration/liquibase/LiquibaseUtils.scala]: Do you really need `mutex`?

It seems like `this.synchronized { ... }` would accomplish the same thing (just as there is only ever one `mutex`, there is only ever one `LiquibaseUtils`)

**Comment by kshakir** [comment]
It will take me a while to dig up an online reference (googling java thread safety returns a ton of results to sort through). But creating one's own private lock is an extra level of paranoia, kind of like marking all java variables as `final`, or reducing the scope of classes to `private`.

If one uses `this` as a mutex, then others can actually steal your lock, by locking **you**.

```scala
object LiquibaseUtils {
  def echoQuick = {
    this.synchronized {
      println("hello");
    }
  }
}

object ForeignImplementation {
  LiquibaseUtils.synchronized {
    // I have your lock!
    Thread.sleep(1.day.toMillis)
  }
}
```

If however the synchronization is done on a private variable, it can never be shared by outside participants.

```scala
object LiquibaseUtils {
  private val cantTouchThis = new Object
  def echoQuick = {
    cantTouchThis.synchronized {
      println("hello");
    }
  }
}

object ForeignImplementation {
  LiquibaseUtils.synchronized {
    // Doesn't affect echoQuick
    Thread.sleep(1.day.toMillis)
  }
}
```

File: database/migration/src/main/scala/cromwell/database/migration/liquibase/LiquibaseUtils.scala (hunk 1 above)
Comment (kshakir) [reply database/migration/src/main/scala/cromwell/database/migration/liquibase/LiquibaseUtils.scala]: See comment below.

File: database/migration/src/main/scala/cromwell/database/migration/liquibase/LiquibaseUtils.scala (hunk 2 above)
Comment (kshakir) [reply database/migration/src/main/scala/cromwell/database/migration/liquibase/LiquibaseUtils.scala]: See comment below.

**Comment by kshakir** [comment]
Found a good discussion. Will put it in a comment.

https://stackoverflow.com/questions/442564/avoid-synchronizedthis-in-java

**Comment by aednichols** [comment]
Cool, TIL

**Pull request merged**